### PR TITLE
Prototype Log4j2 Appender

### PR DESCRIPTION
**Pull request by jack-berg**
*EDIT* - keeping the description below around, but all the issues it mentions have been solved in the 1.9.0 release of the log SDK. This is no longer a prototype. 

This is meant to prompt discussion. Proves out what an `OpenTelemetryAppender` might look like for log4j2. The idea behind this appender is that it delivers log records to a opentelemetry `LogSink`, where they can be batched and shipped out of process via an exporter, such as the recently added OTLP grpc and http log exporters. 

One of the main problems I had to address was that log appenders are instantiated before the opentelemetry log sdk is available. To get around this, each `OpenTelemetryAppender` instance is statically registered after initialization. Once the otel log sdk is available, it is passed to each `OpnTelemetryAppender` instance via a static method, upon which log records start flowing.

Learnings so far:
- LogRecord trace id and span id need to be nullable. And potentially other properties too. The otlp grpc and http exporters currently fail to serialize if span id and trace id are left as their default `""` values. Solved [here](https://github.com/open-telemetry/opentelemetry-java/pull/3745).
- We need a `InMemoryLoggingExporter` for testing. Mocked it out in this code, and will contribute back to `opentelemetry-java`. Solved [here](https://github.com/open-telemetry/opentelemetry-java/pull/3757).
- We need a `SimpleLogProcessor` for testing. Mocked it out in this code, and will contribute back to `opentelemetry-java`. Solved [here](https://github.com/open-telemetry/opentelemetry-java/pull/3749).
- `LogSinkSdkProvider.builder()` is not public but needs to be for it to be useful. I'll make it so. Solved [here](https://github.com/open-telemetry/opentelemetry-java/pull/3745/files).
- `LogSinkSdkProvider` should be immutable after its created. Log processors should be registered with the builder, not with the instance. I'll make it so. Solved [here](https://github.com/open-telemetry/opentelemetry-java/pull/3751).
- Thought needs to be put into how resource and instrumentation library get attached to the log records. Right now and producer to `LogSink` is responsible for creating `LogRecord`s with resource and instrumenation library attached. This means that components like the `OpenTelemetryAppender` need to maintain fields for these. Should `LogSink` accept `LogRecords` without resource and instrumenation library, and let the pipeline attach those?

### CODE REVIEW

File: instrumentation/log4j/log4j-2.13.2/library/src/test/groovy/OpenTelemetryAppenderTest.groovy
Comment (jkwatson) [reply instrumentation/log4j/log4j-2.13.2/library/src/test/groovy/OpenTelemetryAppenderTest.groovy:35]: I think this is a reasonable approach for now. Is it possible to build more complex configuration that would allow log4j to run the initialization itself?

**Comment by jack-berg** [comment]
Now that 1.9.0 has been released, I've updated this branch, refined the logic, and added more thorough tests.

Its now ready for review.

**NOTE** this still places an explicit dependency on the otel java 1.9.0 log SDK. I think the whole repo should be updated to use 1.9.0 before this is merged.

File: instrumentation/log4j/log4j-2.13.2/library/build.gradle.kts
Comment (iNikem) [inline instrumentation/log4j/log4j-2.13.2/library/build.gradle.kts:6]: That will make all clients of this library instrumentation depend on SDK. I assume all users must depend only on API.

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (iNikem) [inline instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:80]:
```suggestion
      builder.setSeverityText(level.name());
```

And shouldn't this text come from our severity and not logging framework level?

File: instrumentation/log4j/log4j-2.13.2/library/build.gradle.kts
Comment (jack-berg) [reply instrumentation/log4j/log4j-2.13.2/library/build.gradle.kts:6]: Users need to call `OpenTelemetryLog4j.initialize(SdkLogEmitterProvider)` to wire opentelemetry into the appender so the log SDK is actually part of this module's API.

Comment (jack-berg) [reply instrumentation/log4j/log4j-2.13.2/library/build.gradle.kts:6]: But this reminds me that I need to update the readme to show usage!

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (jack-berg) [reply instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:80]: From the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-severitytext), the severity text field "is the original string representation of the severity as it is known at the source".

But this does bring up the point that I have made a number of subjective decisions in mapping between the Log4j2 LogEvent data model and the otel log data model. Please take a close look at the attribute name's I've chosen for various things and let me know if you agree / disagree.

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/OpenTelemetryLog4j.java
Comment (mateuszrzeszutek) [inline instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/OpenTelemetryLog4j.java:24]: How is this going to work as a javaagent instrumentation? In the agent we keep all the SDK classes hidden in the agent classloader, they're basically unavailable from the application classloader. The agent won't be able to register its emitterProvider in the application's `OpenTelemetryLog4j` class.
Was the idea about "log appender api" completely scrapped? What happened with that?

File: instrumentation/log4j/log4j-2-common/testing/build.gradle.kts
Comment (anuraaga) [inline instrumentation/log4j/log4j-2-common/testing/build.gradle.kts:8]: Don't think we need version here since we apply BOM for io.opentelemetry. Similar for all usages

File: instrumentation/log4j/log4j-2.13.2/library/build.gradle.kts
Comment (anuraaga) [reply instrumentation/log4j/log4j-2.13.2/library/build.gradle.kts:6]: @iNikem Yeah - we have https://github.com/open-telemetry/opentelemetry-java/issues/3807 to track fixing this.

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (anuraaga) [inline instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:38]: I don't see any semantic conventions for logs defined yet, can we leave them out until they're defined?

https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/logs/semantic_conventions

Comment (anuraaga) [reply instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:38]: If leaving any, I guess logger.name, throwable.name, throwable.message are the only ones I'd keep as they may be necessary for playing with a prototype, but others seem too detailed, or unclear (e.g., I'm not familiar with term `ndc`)

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (anuraaga) [inline instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:53]: Use a method with a `switch` instead of a Map - complete enum switches get compiled into a jump table based on the enum ordinal, which is highly efficient. Errorprone verifying the switch is complete is added icing.

Alternative is `EnumMap`, but we'd only prefer it for an incomplete mapping which doesn't become a jump table.

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (anuraaga) [inline instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:73]: My reading of the javadoc is that `getNanoTime` is just `System.nanoTime()`, not the epoch time

https://logging.apache.org/log4j/2.x/log4j-core/apidocs/org/apache/logging/log4j/core/LogEvent.html#getNanoTime--

`getInstant` is the epoch

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (anuraaga) [inline instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:86]: Hmm - I wonder what this is, I thought the only "class name" type of thing is the logger name, if you derive it from a class. Otherwise each log message would be quite expensive

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (anuraaga) [inline instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:118]: I would like to first try with `Context.current()` and then consider alternatives if we find, e.g. people really want to wrap OTel in a async appenders, it is a significant usability problem.

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (anuraaga) [inline instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:133]: I don't think we should fill arbitrary attributes, I don't think we ever do anything similar for spans

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/OpenTelemetryAppender.java
Comment (anuraaga) [inline instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/OpenTelemetryAppender.java:35]: Since `newBuilder` is public, I'm guessing this class doesn't have to be, but not sure

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/OpenTelemetryLog4j.java
Comment (anuraaga) [inline instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/OpenTelemetryLog4j.java:16]: We generally use `lock` (the idea being that just because it's static final doesn't mean it's constant, as we know a lock has state that mutates)

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/OpenTelemetryLog4j.java
Comment (anuraaga) [reply instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/OpenTelemetryLog4j.java:24]: We'll need it - https://github.com/open-telemetry/opentelemetry-java/issues/3807

File: instrumentation/log4j/log4j-2.13.2/library/src/test/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapperTest.java
Comment (anuraaga) [inline instrumentation/log4j/log4j-2.13.2/library/src/test/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapperTest.java:56]: Better to use `MockitoExtension` and `@Mock` to get strict stubs

Comment (anuraaga) [reply instrumentation/log4j/log4j-2.13.2/library/src/test/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapperTest.java:56]: That being said, I think we'd generally verify mapping via `LogData` instead of using mocks, i.e., these tests could be transfered into the AppenderTest. Imagine adding some helper methods or something on the builder - mocks are nice for RPC stubs, but for data classes we may as well check the actual data.

**Review by anuraaga — COMMENTED** [review]
I think we need a new artifact, maybe `log4j-logssdk-prototype` since we can't leak the SDK dependency into existing users of the log4j instrumentation.

If a temporary artifact in Maven Central is not palatable (it's not great TBH), I think we should hold off for the 1.9.0 release and prioritize https://github.com/open-telemetry/opentelemetry-java/issues/3807. Sorry, didn't notice this ordering constraint until now.

We could still get a snapshot out sooner than later by

- Adding `log4j-logssdk-prototype` after the instrumentation 1.9.0 release, and merging it in for 1.10, using the log appender API
- Adding a `compileOnly` dependency on `log4j` in the opentelemetry-sdk-logs artifact itself and put the code there.

The second option seems to allow the easiest iteration on snapshots.

File: instrumentation/log4j/log4j-2-common/testing/build.gradle.kts
Comment (jack-berg) [reply instrumentation/log4j/log4j-2-common/testing/build.gradle.kts:8]: Yes, I'm just waiting for this project to update its BOM dependency on 1.9.0 and then will drop the version here.

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (jack-berg) [reply instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:38]: ndc = Nested Diagnostic Context

Nested diagnostic context is thread context that acts like a stack. You push items onto it, and pop them off. Mapped diagnostic context is thread context that acts like a map. Frankly, I haven't used Nested Diagnostic Context, and have only used Mapped Diagnostic Context in a limited way in applications I've written.

I took the stance of making sure that all data in the log4j2 data model somehow was mapped into the otel log data model. You're suggesting the opposite, where only a limited part of the log4j2 data model is mapped. I see pros and cons of each stance:
- If we only do a partial mapping of fields well defined in semantic conventions, then we're less likely to make breaking changes, though we can't eliminate breaking changes since there are no stable semantic conventions. We're also going to disappoint folks using parts of the log4j2 data model that are unmapped. 
- If we do a fully mapping, we basically guarantee breaking changes, but at least all the data gets mapped. And by at least making an attempt at mapping each field we can provide better feedback for the semantic conventions.

Personally, I prefer the full mapping approach published in an "alpha" artifact.

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (jack-berg) [reply instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:53]: Wow I didn't know that about enum switch compilation. Cool!

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (jack-berg) [reply instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:73]: Nice catch. Must have missed it because when I tested it with an app on my machine it was returning an epoch nanos.

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (jack-berg) [reply instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:86]: This [test](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4375/files#diff-c3aa4475c1045c52b212b54274b73f836e2cdca46381c0dd2e5601e87243bea9R135) demonstrates the value. In that  case it was `org.apache.logging.log4j.spi.AbstractLogger`. Not sure how useful it is, but as mentioned [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4375#discussion_r749527440), my stance was to map everything.

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (jack-berg) [reply instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:133]: Kind of related to the the conversation about [doing a complete mapping](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4375#discussion_r749527440) or not.

Not filling arbitrary attributes from mapped diagnostic context is limiting the usefulness of the log4j2 api. It means if you want to use otel logs in a more "machine generated events" type of way, that you'd want to use the otel log sdk in your application instead of the log4j2 api.

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (jack-berg) [reply instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:118]: Sounds like a good plan.

File: instrumentation/log4j/log4j-2.13.2/library/src/test/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapperTest.java
Comment (jack-berg) [reply instrumentation/log4j/log4j-2.13.2/library/src/test/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapperTest.java:56]: Yeah looking at this test closer, all the code being exercised is also exercised in `OpenTelemetryAppenderConfigTest`. So I feel comfortable getting rid of this test class altogether.

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/OpenTelemetryLog4j.java
Comment (jack-berg) [reply instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/OpenTelemetryLog4j.java:24]: Took a stab at extracting out the appender API [here](https://github.com/open-telemetry/opentelemetry-java/pull/3880).

File: instrumentation/log4j/log4j-2.13.2/library/build.gradle.kts
Comment (anuraaga) [reply instrumentation/log4j/log4j-2.13.2/library/build.gradle.kts:6]: Sleeping on this, I realized we can keep this code here by making this `compileOnly` - anyone adding the appender has to add the SDK to their classpath themselves.

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (anuraaga) [reply instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:31]: I've seen it's easier to add an attribute and not really go anywhere in the spec with it than the opposite. By not including too much, we can hear about what users really need and then build the spec around it, rather than being ad-hoc. While I don't like to disappoint folks, actually hearing about that disappointment is very valuable feedback.

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (anuraaga) [reply instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:109]: I think we'd generally use baggage for more arbitrary attributes? Though we haven't even specced it for spans. But I'd still suggest keeping things smaller for now to make sure the spec grows rather than going all in

File: instrumentation/log4j/log4j-2.13.2/library/build.gradle.kts
Comment (anuraaga) [reply instrumentation/log4j/log4j-2.13.2/library/build.gradle.kts:6]: Though I still think it might be simpler to continue to prototype with the code in the SDK repo. What do you think?

Comment (jkwatson) [reply instrumentation/log4j/log4j-2.13.2/library/build.gradle.kts:6]: I'm definitely wary about adding an API API to support this. I think `compileOnly` and a reflective check for the presence of an SDK is a better way to go until we have a spec for a new logging API.

Comment (jack-berg) [reply instrumentation/log4j/log4j-2.13.2/library/build.gradle.kts:6]: `compileOnly` dependency on the SDK sounds reasonable to me. 

> Though I still think it might be simpler to continue to prototype with the code in the SDK repo. 

Prototyping in the SDK has the advantage that the SDK and the appender can develop in lockstep, rather than the appender waiting on SDK releases, which is nice. But I think a lot of the SDK flux has been flushed out (of course assuming no significant changes to the experimental log spec) and so I don't expect to see much value from having the appender in the SDK.

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (jack-berg) [reply instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java]: That resonates with me. Also, if we were to do a complete mapping and the required semantic conventions never materialized, we may have to pull support for that mapping, which would be an unpleasant breaking change for users. 

Updated to a minimalist mapping in new commit.

File: instrumentation/log4j/log4j-2.13.2/library/build.gradle.kts
Comment (anuraaga) [inline instrumentation/log4j/log4j-2.13.2/library/build.gradle.kts:6]: The repo's updated to 1.9 so should be able to merge main and remove the versions now

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java
Comment (anuraaga) [inline instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/LogEventMapper.java:98]: While log4j probably doesn't ever add log levels, it's conceivable and could trigger this line if log4j is newer than what we've gotten to compile against. Let's return UNDEFINED_SEVERITY_NUMBER instead of throwing

**Review by anuraaga — APPROVED** [review]

**Review by mateuszrzeszutek — APPROVED** [review]

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/OpenTelemetryAppender.java
Comment (trask) [inline instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/OpenTelemetryAppender.java:31]: we've recently standardized on `builder` over `newBuilder` (#4392)
```suggestion
  public static <B extends Builder<B>> B builder() {
```

File: instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/OpenTelemetryLog4j.java
Comment (trask) [inline instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/OpenTelemetryLog4j.java:28]: given trickiness around debugging init sequences, is it worth adding the stack trace of the original caller similar to GlobalOpenTelemetry?

https://github.com/open-telemetry/opentelemetry-java/blob/efbc87b4df38ea457c159b9360ae0db8a1791bb1/api/all/src/main/java/io/opentelemetry/api/GlobalOpenTelemetry.java#L82-L93

Comment (jack-berg) [reply instrumentation/log4j/log4j-2.13.2/library/src/main/java/io/opentelemetry/instrumentation/log4j/v2_13_2/OpenTelemetryLog4j.java:28]: I like that idea.

**Comment by jack-berg** [comment]
Can this PR be merged?

**Comment by trask** [comment]
🎉

**Pull request merged**
